homeassistant/components/zimi/config_flow.py

```diff
@@ -0,0 +1,142 @@
+"""Config flow for zcc integration."""
+
+from __future__ import annotations
+
+import logging
+import socket
+from typing import Any
+
+import voluptuous as vol
+from zcc import ControlPointDiscoveryService, ControlPointError
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_HOST, CONF_PORT
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.device_registry import format_mac
+
+from .const import DOMAIN, MAC, TIMEOUT, VERBOSITY, WATCHDOG
+
+_LOGGER = logging.getLogger(__name__)
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_HOST, default=""): str,
+        vol.Optional(CONF_PORT, default=5003): int,
+        vol.Required(MAC): str,
+    }
+)
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for zcc."""
+
+    VERSION = 1
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> config_entries.ConfigFlowResult:
+        """Handle the initial step."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=STEP_USER_DATA_SCHEMA
+            )
+
+        data = {}
+        errors = {}
+
+        try:
+            data = await self.validate_input(user_input)
+        except CannotConnect:
+            errors["base"] = "cannot_connect"
+        except ConnectionRefused:
+            errors["base"] = "connection_refused"
+        except DiscoveryFailure:
+            errors["base"] = "discovery_failure"
+        except InvalidAuth:
+            errors["base"] = "invalid_auth"
+        except InvalidHost:
+            errors["base"] = "invalid_host"
+        except TimeOut:
+            errors["base"] = "timeout"
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception during configuration steps")
+            errors["base"] = "unknown"
+        else:
+            await self.async_set_unique_id(data["mac"])
+            return self.async_create_entry(title=data["title"], data=data)
+
+        return self.async_show_form(
+            step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
+        )
+
+    async def validate_input(self, data: dict[str, Any]) -> dict[str, Any]:
```

Comment:
will you reuse that somewhere?

Reply:
To the validate function. If it is only used once, it's better to inline this code imo

Reply:
Refactored with inline, better descriptions, inverted flow. See https://github.com/home-assistant/core/commit/7998adc1c0be9af723bc19f92660a8976a33db7a